### DISCUSSION TITLE
How to unit test popper virtual elements in Jest?

### DISCUSSION THREAD

**Discussion by JJ-Kamminga** [floating-ui/floating-ui#1383] [Q&A]
How to unit test popper virtual elements in Jest (I use React Testing Library)?
I use react-popper, but I suppose this question applies to popper generally.

I used to do this:
```
const virtualReference = {
  getBoundingClientRect(): DOMRect {
    return new DOMRectReadOnly(200, 10, 90, 10);
  },
};
```
And then pass this to `usePopper()`.

But since I upgraded Node (now on 16.10), my Jest call retries get exceeded on this error:
`UnhandledPromiseRejectionWarning: ReferenceError: DOMRectReadOnly is not defined`

I suspect the problem is with my tests, not with Node.
What do you think is the right way to test this? Mock the virtual element? If so, how? Is this even something that unit tests should cover?

**Comment by FezVrasta** [comment]
I never used DOMRectReadonly, I usually pass an object with `top`, `left`, `right`, `bottom`, `height` and `width` properties.

**Reply by JJ-Kamminga** [reply]
Do you use TypeScript? Curious how you would type it.

**Reply by FezVrasta** [reply]
I use Flow, I never had to type it because Flow auto inferred the right types.
